File: 414.第三大的数.py
#
# @lc app=leetcode.cn id=414 lang=python
#
# [414] 第三大的数
#

# @lc code=start
class Solution(object):
    def thirdMax(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        first = second = third = float('-inf')
        for num in nums:
            if num > third:
                if num < second:
                    third = num
                elif num > second:
                    if num < first:
                        third = second
                        second = num
                    elif num > first:
                        third = second
                        second = first
                        first = num
        if third == float('-inf'):
            return first
        else:
            return third
        
# @lc code=end

File: 414.第三大的数.cpp
/*
 * @lc app=leetcode.cn id=414 lang=cpp
 *
 * [414] 第三大的数
 */

// @lc code=start
class Solution {
public:
    int thirdMax(vector<int>& nums) {
        sort(nums.begin(), nums.end());
        if(nums.size() <3)
            return nums[nums.size() -1];
        int maxcount = 2;
        int val = nums[nums.size() -1];
        for(int i = nums.size()-2; i >= 0; i--){
            if(val != nums[i])
            {
                maxcount--;
                val = nums[i];
                if(maxcount == 0)
                    return val;
            }
        }
        if(maxcount)
            return nums[nums.size() -1];
        return val;

    }
};
// @lc code=end

